### DISCUSSION TITLE
LED brightness control

### DISCUSSION THREAD

**Discussion by XJIOP** [Adam777Z/xbox-controller-button-remapper#16] [General]
Is it possible to add LED brightness control? I know this can be done when Xbox Controller is connected by wire or Xbox Wireless Adapter except Bluetooth.

What I found:
- LED can be controlled with cmd SETX SDL_JOYSTICK_HIDAPI_XBOX_ONE_HOME_LED 0.05 then start Steam.
- Paid app reWASD can do it.
- https://github.com/JRHeaton/XboxOneControllerDev

**Comment by Adam777Z** [comment]
I thought about this before (months ago) and explored it further then, not long after this commit:
[Added the hint SDL_HINT_JOYSTICK_HIDAPI_XBOX_ONE_HOME_LED to control the brightness of the Xbox button LED on the Xbox One controller (libsdl-org/SDL@24cdebe)](https://github.com/libsdl-org/SDL/commit/24cdebe464342e83743eaa613efacab4ef136c58)

I tried it and it did not work so I contacted the creator of SDL (@slouken) to find out why and got the answer that it requires a driver on Windows to work. That is how Steam and reWASD work.

There could be a simple program (just like this one) that comes with a driver and controls only the Xbox Controller's Xbox button light/LED that is also open source and free (just like this one). I am currently not aware of such program. It may or may not exist (yet) and may need to be created.

There is a free solution with Steam, it looks like the best free method at this time:
[Any NEW solution for the Xbox button Light? (controller) - Steam Forums](https://steamcommunity.com/discussions/forum/1/3819656548992688579/#c3819656548994120770)

**Reply by XJIOP** [reply]
Perhaps this can be done through xinput.dll?
https://github.com/metarutaiga/DXSDK/blob/master/XInput.h

**Reply by Adam777Z** [reply]
No, currently a driver is required on Windows.

**Comment by Adam777Z** [comment]
It should be possible to change the Xbox Controller's Xbox button light/LED brightness using the official [Xbox Accessories app](https://support.xbox.com/en-US/help/hardware-network/accessories/xbox-accessories-app). Hopefully it will be added soon.

**Reply by Aragorn-E** [reply]
> It should be possible to change the Xbox Controller's Xbox button light/LED brightness using the official [Xbox Accessories app](https://support.xbox.com/en-US/help/hardware-network/accessories/xbox-accessories-app). Hopefully it will be added soon.

No, it's not possible with the Xbox Accessories app, and I'm over 2 years in the future.

**Reply by Leclowndu93150** [reply]
hello i created this https://github.com/Leclowndu93150/xbledctl

**Reply by Adam777Z** [reply]
Thank you for this! I tried it and it works perfectly. Finally a standalone, simple program to control the Xbox Controller's Xbox button light/LED! There is no longer a need for the Steam method.

**Reply by Leclowndu93150** [reply]
took me a few hours of research to figure out haha, you can read my research file if you are technical enough

**Reply by Adam777Z** [reply]
I already did. Great work!

**Reply by Aragorn-E** [reply]
> hello i created this https://github.com/Leclowndu93150/xbledctl

Thanks. It works well. I just wish there was a solution for wireless, as that's when turning off LED becomes useful; to preserve battery charge.

I use a Wifi USB Dongle for my controller. Does that count for Bluetooth? As to my understanding, your tool doesn't work for Bluetooth.

**Reply by Leclowndu93150** [reply]
> > hello i created this https://github.com/Leclowndu93150/xbledctl
> 
> Thanks. It works well. I just wish there was a solution for wireless, as that's when turning off LED becomes useful; to preserve battery charge.
> 
> I use a Wifi USB Dongle for my controller. Does that count for Bluetooth? As to my understanding, your tool doesn't work for Bluetooth.


Unfortunately there will never be support for bluetooth as it is just impossible but i think the microsoft adapter works

**Reply by Aragorn-E** [reply]
> > > hello i created this https://github.com/Leclowndu93150/xbledctl
> > 
> > 
> > Thanks. It works well. I just wish there was a solution for wireless, as that's when turning off LED becomes useful; to preserve battery charge.
> > I use a Wifi USB Dongle for my controller. Does that count for Bluetooth? As to my understanding, your tool doesn't work for Bluetooth.
> 
> Unfortunately there will never be support for bluetooth as it is just impossible but i think the microsoft adapter works

I understand. If it changes anything (I doubt it), the USB Dongle I'm using is called "GuliKit Hyperlink 2".

**Reply by XJIOP** [reply]
@Leclowndu93150 I confirm that it works with the Microsoft adapter. Are you planning a stylish tray icon?

**Reply by Leclowndu93150** [reply]
> @Leclowndu93150 I confirm that it works with the Microsoft adapter. Are you planning a stylish tray icon?

If you get me one :)

**Reply by Adam777Z** [reply]
> I understand. If it changes anything (I doubt it), the USB Dongle I'm using is called "GuliKit Hyperlink 2".

That uses Bluetooth.

It works with the [Xbox Series X|S Controller](https://en.wikipedia.org/wiki/Xbox_Wireless_Controller#Third_revision_(2020)) wired and wireless using the [Xbox Wireless Adapter](https://en.wikipedia.org/wiki/Xbox_Wireless_Controller#Xbox_Wireless_Adapter_for_Windows).

**Reply by Aragorn-E** [reply]
> > I understand. If it changes anything (I doubt it), the USB Dongle I'm using is called "GuliKit Hyperlink 2".
> 
> That uses Bluetooth.
> 
> It works with the [Xbox Series X|S Controller](https://en.wikipedia.org/wiki/Xbox_Wireless_Controller#Third_revision_(2020)) wired and wireless using the [Xbox Wireless Adapter](https://en.wikipedia.org/wiki/Xbox_Wireless_Controller#Xbox_Wireless_Adapter_for_Windows).

Thanks for the info. If only I could find where to buy that adapter... Only ones I found were knock off copycats pretending to be the official one.

**Reply by Leclowndu93150** [reply]
https://www.amazon.fr/Adaptateur-Xbox-sans-pour-Windows/dp/B074QMN4YW?source=ps-sl-shoppingads-lpcontext&psc=1&smid=AA4CZO0WP0RGS mb for the french

**Reply by XJIOP** [reply]
I had to buy a second Xbox One controller that came with this USB adapter because it was cheaper than buying it separately.
